doc/developer/design/20230829_topk_size_hint.md

```diff
@@ -0,0 +1,426 @@
+FROM (
+    SELECT DISTINCT ON(teacher_id) id, teacher_id, MAX(course_id) AS max_course_id
+    FROM sections
+    GROUP BY id, teacher_id
+    OPTIONS (EXPECTED GROUP SIZE = 1000)
+    ORDER BY teacher_id, id
+    LIMIT 2
+);
+
+EXPLAIN PLAN FOR MATERIALIZED VIEW nested_distinct_on_group_by_limit;
+                                      Optimized Plan
+------------------------------------------------------------------------------------------
+ materialize.public.nested_distinct_on_group_by_limit:                                   +
+   Return                                                                                +
+     Union                                                                               +
+       Get l0                                                                            +
+       Map (null, null, null)                                                            +
+         Union                                                                           +
+           Negate                                                                        +
+             Project ()                                                                  +
+               Get l0                                                                    +
+           Constant                                                                      +
+             - ()                                                                        +
+   With                                                                                  +
+     cte l0 =                                                                            +
+       Reduce aggregates=[sum(#0), sum(#1), sum(#2)]                                     +
+         TopK order_by=[#1 asc nulls_last, #0 asc nulls_last] limit=2 exp_group_size=1000+
+           TopK group_by=[#1] order_by=[#0 asc nulls_last] limit=1 exp_group_size=1000   +
+             Reduce group_by=[#0, #1] aggregates=[max(#2)] exp_group_size=1000           +
+               Project (#0..=#2)                                                         +
+                 Get materialize.public.sections                                         +
+
+(1 row)
+```
+
+To disambiguate the query hints when necessary, we argue for an approach with the
+following characteristics:
+1. Maintain backwards compatibility with `EXPECTED GROUP SIZE` whenever there is no ambiguity, i.e.,
+if only the `EXPECTED GROUP SIZE` is specified, it is attached to all instances of reductions and
+top-k operators.
+2. When there is ambiguity, allow for the specification of additional query hints that attach
+to specific clauses in the query block, overriding the `EXPECTED GROUP SIZE` (if any is given).
+
+To operationalize the above, two additional disambiguation hints are proposed:
+
+1. `LIMIT GROUP SIZE`: This hint attaches to the `TopK` operator implementing the `LIMIT` clause.
```

Comment:
If we are reserving `EXPECTED GROUP SIZE` for some smart behavior (figure out which operator to target based on the plan), I suggest to add explicitly `AGGREGATE GROUP SIZE` or `REDUCE GROUP SIZE` to also precisely tackle the `Reduce` operator resulting from planning a `GROUP BY` clause. Without a dedicated option, it seems to me that the semantics of `EXPECTED GROUP SIZE` will become too wild:

1. In the presence of other options, it means specifically targeting the `GROUP BY` operator.
2. In the absence of other options, it means targeting the only `Reduce` or `TopK` in the enclosing syntactic scope.

Reply:
I think here the semantics for `EXPECTED GROUP SIZE` would be slightly different:
1. In the absence of other options, the `EXPECTED GROUP SIZE` applies to *all* `Reduce` and `TopK` instances in the enclosing syntactic scope.
2. In the presence of other options, the other options override the `EXPECTED GROUP SIZE` for the `TopK` instances in the specific clauses targeted.

Reply:
The semantics above are backwards compatible, which is an important design goal to control operational impacts here. It is still possible to add an `AGGREGATE GROUP SIZE` that only targets the corresponding `Reduce` operator, but I am not sure if the added complexity is worth it. Namely, right now, we could explain to users that essentially they can use `EXPECTED GROUP SIZE` throughout and if there is ambiguity, rely on the `TopK` hints to break it. With `AGGREGATE GROUP SIZE`, it is odd to explain why we'd be keeping `EXPECTED GROUP SIZE` around.

Reply:
> it is odd to explain why we'd be keeping `EXPECTED GROUP SIZE` around.

I was thinking that we can just remove mentions of `EXPECTED GROUP SIZE` from the official docs and keep it around for backward compatibility.

Reply:
My concern is that maintaining undocumented options around would give us operational headaches in the future. For example, what is the semantics of queries that specify both `EXPECTED GROUP SIZE` and some of the new options? Users who know about the old hint could still use it in confusing ways together with the new ones.

And if we ever wanted to get rid of the old hint, we'd need to make sure to either write migrations (with operational risks that mistakes could blow up memory for clusters that are already running) or track down users to rewrite their queries to use the new hints (e.g., through a syntax deprecation process).

It sounds like quite a bit of work and/or risk for a change in syntax that I am not sure would be buying us a lot. Even today, users could deal with the ambiguity by just breaking down their queries into multiple query blocks as a workaround.

Reply:
> For example, what is the semantics of queries that specify both EXPECTED GROUP SIZE and some of the new options?

They should not exist. They don't exist today and we can write the new code to don't allow such combinations.

> And if we ever wanted to get rid of the old hint, we'd need to make sure to either write migrations (with operational risks that mistakes could blow up memory for clusters that are already running) or track down users to rewrite their queries to use the new hints (e.g., through a syntax deprecation process).

I don't think we should be afraid of that, because we are not forced to migrate.

To illustrate my concerns, let me give you an example:

Say I write a query:

```sql
SELECT
    sum(max_revenue) AS sum_max_revenue
FROM (
    SELECT l_orderkey,
           max(l_extendedprice * (1 - l_discount)) AS max_revenue
    FROM lineitem
    GROUP BY l_orderkey
    ORDER BY max_revenue DESC
    LIMIT 10
);
```

and then I decide to give a hint to both the `LIMIT` and the `GROUP BY`:

```sql
OPTIONS (EXPECTED GROUP SIZE = 7, LIMIT GROUP SIZE = 1000000)
```

Life is good - `Reduce` gets `exp_group_size=7` and `TopK`  gets `exp_group_size = 1000000`. Months or years pass, and after a while some other person looks at the query and decides that the `LIMIT GROUP SIZE` is probably off. Rather than doing the work to figure and set the right value, they just remove `LIMIT GROUP SIZE` altogether! Suddenly, instead of removing the `TopK` hint altogether, we start inheriting the one from `EXPECTED GROUP SIZE`.

I don't think such experiences help with establishing trust.

Reply:
>> For example, what is the semantics of queries that specify both EXPECTED GROUP SIZE and some of the new options?

> They should not exist. They don't exist today and we can write the new code to don't allow such combinations.

It is not very standard for hint misconfiguration to error out queries (most DBMS just ignore incorrect hinting). But I would not be against having us explicitly produce errors since hints are part of a query clause in the case of Materialized (as opposed to some comment-like syntax in other DBMS).

I will revise the document then to introduce the `AGGREGATE GROUP SIZE`, include the error semantics on mixed hinting, and the preservation of the `EXPECTED GROUP SIZE` as it is today.

Reply:
> I will revise the document then to introduce the AGGREGATE GROUP SIZE, include the error semantics on mixed hinting, and the preservation of the EXPECTED GROUP SIZE as it is today.

Thank you. I understand that there is no perfect solution, and if other people think that I'm overestimating the risk [from customers hitting the above example](https://github.com/MaterializeInc/materialize/pull/21500#discussion_r1311843825), I'm happy to proceed with the original design.

> But I would not be against having us explicitly produce errors since hints are part of a query clause in the case of Materialized (as opposed to some comment-like syntax in other DBMS).

Yes, I guess we are lucky in that regard. TBH I suspect we wouldn't be having this discussion if we had comment-level hints because the comments would appear just before or after the syntax which they want to target.

Reply:
My thinking is also along the lines of what Alex said. If we didn't have to worry about backwards compatibility, we would just add the specific query hints and drop `EXPECTED GROUP SIZE` entirely because it is confusing. The reality is that backwards compatibility is important, so we can't do the last part. But we can deprecate `EXPECTED GROUP SIZE`. I don't think we have processes for deprecating syntax today, but there are a couple things we could do:

1. Just don't document `EXPECTED GROUP SIZE` and hope that people eventually forget about it.
2. Print a deprecation warning/notice when people send queries that use `EXPECTED GROUP SIZE`.

I'd think (1) alone is probably sufficient for now, given how little used this hint probably is today. In a year or so we can check all the prod catalogs to see if anyone still uses `EXPECTED GROUP SIZE` or not.

Reply:
I updated the document to reflect this discussion now, so it might be worth a look at the second commit with the changes.